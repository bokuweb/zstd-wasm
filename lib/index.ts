const Module = require("./wasm/zstd.js");

const initialized = (() =>
  new Promise<void>((resolve) => {
    Module.onRuntimeInitialized = resolve;
  }))();

const isError = (code: number): number => {
  const _isError = Module.cwrap("ZSTD_isError", "number", ["number"]);
  return _isError(code);
};

const getErrorName = (code: number): string => {
  const _getErrorName = Module.cwrap("ZSTD_getErrorName", "string", ["number"]);
  return _getErrorName(code);
};

const getFrameContentSize = (buf: ArrayBuffer): number => {
  const getSize = Module.cwrap("ZSTD_getFrameContentSize", "number", [
    "array",
    "number",
  ]);
  return getSize(buf, buf.byteLength);
};

export const decompress = async (buf: ArrayBuffer): Promise<ArrayBuffer> => {
  await initialized;
  const contentSize = getFrameContentSize(buf);
  const size = contentSize === -1 ? 1024 * 1024 : contentSize;
  const malloc = Module.cwrap("create_buffer", "number", ["number"]);
  const free = Module.cwrap("destroy_buffer", "number");
  const heap = malloc(size);
  let code;
  try {
    const _decompress = Module.cwrap("ZSTD_decompress", "number", [
      "number",
      "number",
      "array",
      "number",
    ]);
    code = _decompress(heap, size, buf, buf.byteLength);
    if (isError(code)) {
      throw new Error(getErrorName(code));
    }
  } catch (e) {
    free(heap);
    throw e;
  }
  free(heap);
  return new Uint8Array(Module.HEAPU8.buffer, heap, code);
};
