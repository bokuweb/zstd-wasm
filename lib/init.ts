const Module = require('./wasm/zstd.js');

const initialized = (() =>
  new Promise<void>((resolve) => {
    Module.onRuntimeInitialized = resolve;
  }))();

export const waitInitialized = async () => {
  await initialized;
};

export const init = async (/*path?: string*/) => {
  // Module['init'](path);
  await waitInitialized();
};

export { Module };
