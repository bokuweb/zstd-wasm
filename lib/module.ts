import { Module } from './zstd';

const initialized = (() =>
  new Promise<void>((resolve) => {
    Module.onRuntimeInitialized = resolve;
  }))();

export const waitInitialized = async () => {
  await initialized;
};

export { Module };
