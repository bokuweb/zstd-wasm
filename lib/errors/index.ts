import { Module } from '../init';

export const isError = (code: number): number => {
  const _isError = Module.cwrap('ZSTD_isError', 'number', ['number']);
  return _isError(code);
};

// @See https://github.com/facebook/zstd/blob/12c045f74d922dc934c168f6e1581d72df983388/lib/common/error_private.c#L24-L53
export const getErrorName = (code: number): string => {
  const _getErrorName = Module.cwrap('ZSTD_getErrorName', 'string', ['number']);
  return _getErrorName(code);
};
