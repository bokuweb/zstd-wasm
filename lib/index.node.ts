import { Module, waitInitialized } from './module';

export const init = async () => {
  const { readFile } = require('fs/promises');
  const { resolve } = require('path');
  const buf = await readFile(resolve(__dirname, './zstd.wasm'));
  Module['init'](buf);
  await waitInitialized();
};

export * from './simple/decompress';
export * from './simple/compress';

export * from './simple/decompress_using_dict';
export * from './simple/compress_using_dict';
