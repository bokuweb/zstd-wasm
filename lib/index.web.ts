import { Module, waitInitialized } from './module';

export const init = async (path?: string) => {
  const url = require('./wasm/zstd.wasm');
  Module['init'](path ?? url.default ?? url);
  await waitInitialized();
};

export * from './simple/decompress';
export * from './simple/compress';

export * from './simple/decompress_using_dict';
export * from './simple/compress_using_dict';
