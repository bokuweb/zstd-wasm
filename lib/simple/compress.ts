import { waitInitialized, Module } from '../init';
import { isError, getErrorName } from '../errors';

const compressBound = (size: number): number => {
  const bound = Module.cwrap('ZSTD_compressBound', 'number', ['number']);
  return bound(size);
};

export const compress = async (buf: ArrayBuffer, level: number) => {
  await waitInitialized();
  const bound = compressBound(buf.byteLength);
  const malloc = Module.cwrap('create_buffer', 'number', ['number']);
  const compressed = malloc(bound);
  const free = Module.cwrap('destroy_buffer', 'number');
  try {
    /*
      @See https://zstd.docsforge.com/dev/api/ZSTD_compress/
      size_t ZSTD_compress( void* dst, size_t dstCapacity, const void* src, size_t srcSize, int compressionLevel);
      Compresses `src` content as a single zstd compressed frame into already allocated `dst`.
      Hint : compression runs faster if `dstCapacity` >=  `ZSTD_compressBound(srcSize)`.
      @return : compressed size written into `dst` (<= `dstCapacity),
                or an error code if it fails (which can be tested using ZSTD_isError()).
    */
    const _compress = Module.cwrap('ZSTD_compress', 'number', ['number', 'number', 'array', 'number', 'number']);
    const sizeOrError = _compress(compressed, bound, buf, buf.byteLength, level);
    if (isError(sizeOrError)) {
      throw new Error(getErrorName(sizeOrError));
    }
    // // Copy buffer
    // // Uint8Array.prototype.slice() return copied buffer.
    const data = new Uint8Array(Module.HEAPU8.buffer, compressed, sizeOrError).slice();
    free(compressed, bound);
    return data;
  } catch (e) {
    free(compressed, bound);
    throw e;
  }
};
