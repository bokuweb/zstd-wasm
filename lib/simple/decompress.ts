import { Module } from '../module';
import { isError } from '../errors';

const getFrameContentSize = (src: number, size: number): number => {
  const getSize = Module['_ZSTD_getFrameContentSize'];
  return getSize(src, size);
};

export type DecompressOption = {
  defaultHeapSize?: number;
};

export const decompress = (
  buf: Uint8Array,
  opts: DecompressOption = { defaultHeapSize: 1024 * 1024 }, // Use 1MB on default if it is failed to get content size.
): Uint8Array => {
  const malloc = Module['_malloc'];
  const src = malloc(buf.byteLength);
  Module.HEAP8.set(buf, src);
  const contentSize = getFrameContentSize(src, buf.byteLength);
  const size = contentSize === -1 ? opts.defaultHeapSize : contentSize;
  const free = Module['_free'];
  const heap = malloc(size);
  try {
    /*
      @See https://zstd.docsforge.com/dev/api/ZSTD_decompress/
      compressedSize : must be the exact size of some number of compressed and/or skippable frames.
      dstCapacity is an upper bound of originalSize to regenerate.
      If user cannot imply a maximum upper bound, it's better to use streaming mode to decompress data.
      @return: the number of bytes decompressed into dst (<= dstCapacity), or an errorCode if it fails (which can be tested using ZSTD_isError()).
    */
    const _decompress = Module['_ZSTD_decompress'];
    const sizeOrError = _decompress(heap, size, src, buf.byteLength);
    if (isError(sizeOrError)) {
      throw new Error(`Failed to compress with code ${sizeOrError}`);
    }
    // Copy buffer
    // Uint8Array.prototype.slice() return copied buffer.
    const data = new Uint8Array(Module.HEAPU8.buffer, heap, sizeOrError).slice();
    free(heap, size);
    free(src, buf.byteLength);
    return data;
  } catch (e) {
    free(heap, size);
    free(src, buf.byteLength);
    throw e;
  }
};
