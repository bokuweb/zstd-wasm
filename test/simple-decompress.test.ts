import { decompress } from '../lib/simple/decompress';
import { init } from '../lib/init';

test('decompress multiple json', async () => {
  const dataA =
    'KLUv/QBgpQQA0gkfHWClqgMIrrCtT3yQwJQIidYIgtCCI8CfnRdBEHABc7LJJpvs0EdlsMvRV0fXLl7Re/k5y+/oEoAlM8m7ngUB8uhKgw859JEdZI7ml9FqWzVhdafmiE1CE0MyWhmpWEOh1lBc75I/+pXrjvSS4aVp0B2dm7fO9bJxAIF/AgcARQ3FJ+JWVRjGDm8FszjLChgqDw==';

  const dataB =
    'KLUv/QBgnQQAwskeHWClqgMIrrCtT3yQwJQIidYIgtCCI8CfnRdBEHABc7LJJpvs0EdlsMvRO7p28Yrey89ZfkeXACyZSd71LAiQR1cafMihj+wgczS/jFbbqgmrOzVHbBKaGJLRykjFGgq1huJ6l/zRr1x3pJcML02D7ujcvHWul40DCPwTBwBFDcUn4lZVGMYObwWzOMsKGCoP';

  await init();
  const resA = decompress(Buffer.from(dataA, 'base64'));
  const jsonA = Buffer.from(resA).toString();
  expect(JSON.parse(jsonA)).toMatchSnapshot();

  const resB = decompress(Buffer.from(dataB, 'base64'));
  const jsonB = Buffer.from(resB).toString();
  expect(JSON.parse(jsonB)).toMatchSnapshot();
});
