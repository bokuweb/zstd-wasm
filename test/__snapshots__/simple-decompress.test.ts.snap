// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decompress multiple json 1`] = `
Object {
  "document": Object {
    "data": Object {
      "id": "74207427-90f9-46f1-a6dc-ff21c0fa0f92",
    },
    "nodes": Array [
      Object {
        "data": Object {},
        "nodes": Array [
          Object {
            "leaves": Array [
              Object {
                "marks": Array [],
                "object": "leaf",
                "text": "sdfdsf2",
              },
            ],
            "object": "text",
          },
        ],
        "object": "block",
        "type": "line",
      },
    ],
    "object": "document",
  },
  "object": "value",
}
`;

exports[`decompress multiple json 2`] = `
Object {
  "document": Object {
    "data": Object {
      "id": "74207427-90f9-46f1-a6dc-ff21c0fa0f92",
    },
    "nodes": Array [
      Object {
        "data": Object {},
        "nodes": Array [
          Object {
            "leaves": Array [
              Object {
                "marks": Array [],
                "object": "leaf",
                "text": "sdfdsf",
              },
            ],
            "object": "text",
          },
        ],
        "object": "block",
        "type": "line",
      },
    ],
    "object": "document",
  },
  "object": "value",
}
`;
